### ISSUE TITLE
the result  in the test phase （main_nyu_posereg_embedding）

### ISSUE THREAD

**Issue by sinewy333** [moberweger/deep-prior#21]
I was using the main_nyu_posereg_embedding code in tensorflow, The output is:
Training
epoch 100, batch_num 1135, Minibatch Loss= 1.9201
Testing ...
Mean error: 483.2774520085452mm, max error: 682.6990653506009mm
Testing baseline
Mean error: 33.98014831542969mm
Mean error: 615.936767578125mm
I've tried to change all of the parameters, but I can't reduce the error value.This problem has been bothering me for a long time.So I hope you can tell me,Is the result correct? thank you  very much!

**Comment by moberweger** [comment]
Hi,
I am not sure what you changed in order to run the code in tensorflow,
since it is originally written for theano. But as far as I can see from the
information you mentioned, there are at least two problems:
- the network loss looks like it did not converge properly. Usually it is
around 10e-3, but for you it is 1.9.
- there is a problem with the data, since testing the baseline gives you
600mm error on the second testset, which is too large. I suggest validating
and debugging the data that you feed to the network.

**Comment by sinewy333** [comment]
my code is：
import numpy
import matplotlib
matplotlib.use('Agg')  # plot to file
import matplotlib.pyplot as plt
import os
from sklearn.externals import joblib
import sys
sys.path.append("E://data")
from sklearn.decomposition import PCA
from data.importers import NYUImporter
from data.dataset import NYUDataset
from util.handpose_evaluation import NYUHandposeEvaluation
from data.transformations import transformPoint2D
import cv2
import tensorflow as tf



eval_prefix = 'NYU_EMB_t0nF8mp421fD553h1024_PCA30'
if not os.path.exists('./eval/'+eval_prefix+'/'):
  os.makedirs('./eval/'+eval_prefix+'/')


rng = numpy.random.RandomState(23455)

print("create data")

di = NYUImporter('../NYU/')
Seq1 = di.loadSequence('train', shuffle=True, rng=rng)
trainSeqs = [Seq1]

Seq2_1 = di.loadSequence('test_1')
Seq2_2 = di.loadSequence('test_2')
testSeqs = [Seq2_1, Seq2_2]

    # create training data
trainDataSet = NYUDataset(trainSeqs)
train_data, train_gt3D = trainDataSet.imgStackDepthOnly('train')

mb = (train_data.nbytes) / (1024 * 1024)
print("data size: {}Mb".format(mb))

valDataSet = NYUDataset(testSeqs)
val_data, val_gt3D = valDataSet.imgStackDepthOnly('test_1')

testDataSet = NYUDataset(testSeqs)
test_data1, test_gt3D1 = testDataSet.imgStackDepthOnly('test_1')
test_data2, test_gt3D2 = testDataSet.imgStackDepthOnly('test_2')

print(train_gt3D.max(), test_gt3D1.max(), train_gt3D.min(), test_gt3D1.min())
print(train_data.max(), test_data1.max(), train_data.min(), test_data1.min())

imgSizeW = train_data.shape[3]
imgSizeH = train_data.shape[2]
nChannels = train_data.shape[1]

    ####################################
    # convert data to embedding
pca = PCA(n_components=30)
pca.fit(train_gt3D.reshape((train_gt3D.shape[0], train_gt3D.shape[1]*3)))
train_gt3D_embed = pca.transform(train_gt3D.reshape((train_gt3D.shape[0], train_gt3D.shape[1]*3)))
test_gt3D_embed1 = pca.transform(test_gt3D1.reshape((test_gt3D1.shape[0], test_gt3D1.shape[1]*3)))
test_gt3D_embed2 = pca.transform(test_gt3D2.reshape((test_gt3D2.shape[0], test_gt3D2.shape[1]*3)))
val_gt3D_embed = pca.transform(val_gt3D.reshape((val_gt3D.shape[0], val_gt3D.shape[1]*3)))
    
X0 = tf.placeholder(tf.float32, shape=(None, 1, 128, 128))
Y = tf.placeholder(tf.float32, shape=(None, 30))

    ############################################################################
    
def conv2d(x, W, b):
    x = tf.nn.conv2d(x, W, strides=[1, 1, 1, 1], padding='SAME')
    x = tf.nn.bias_add(x, b)
    return tf.nn.relu(x)		

def maxpool2d(x, k):
    return tf.nn.max_pool(
        x,
        ksize = [1, k, k, 1],
        strides = [1, k, k, 1],
        padding='SAME')

def create_model(x0, weights, biases):
    
    x0 = tf.reshape(x0, shape=[-1, 128, 128, 1])
    conv1_1 = conv2d(x0, weights['wc1_1'], biases['bc1_1'])
    conv1_1 = tf.nn.relu(maxpool2d(conv1_1, k=4))
    conv1_2 = conv2d(conv1_1, weights['wc1_2'], biases['bc1_2'])
    conv1_2 = tf.nn.relu(maxpool2d(conv1_2, k=2))
    conv1_3 = conv2d(conv1_2, weights['wc1_3'], biases['bc1_3'])
    conv1_3 = tf.nn.relu(maxpool2d(conv1_3, k=1))
    
    fc1 = tf.contrib.layers.flatten(conv1_3)
    fc1 = tf.add(tf.matmul(fc1, weights['fc1']), biases['fc1'])
    fc1 = tf.nn.relu(fc1)
    fc1 = tf.nn.dropout(fc1,0.7,noise_shape=None,seed=None,name=None)
    fc2 = tf.add(tf.matmul(fc1, weights['fc2']), biases['fc2'])
    fc2 = tf.nn.relu(fc2)
    fc2 = tf.nn.dropout(fc2,0.7,noise_shape=None,seed=None,name=None) 
    out = tf.add(tf.matmul(fc2, weights['out']), biases['out'])
    return out


scale1 = numpy.sqrt(2. / 25.)
scale2 = numpy.sqrt(2. / 25.) 
scale3 = numpy.sqrt(2. / 9.) 
                                         
weights = {
'wc1_1': tf.Variable(tf.random_normal(shape=[5, 5, 1, 8],mean=0.0,stddev=scale1),name='w1'),
'wc1_2': tf.Variable(tf.random_normal(shape=[5, 5, 8, 8],mean=0.0,stddev=scale2),name='w2'),
'wc1_3': tf.Variable(tf.random_normal(shape=[3, 3, 8, 8],mean=0.0,stddev=scale3),name='w3'),
'fc1': tf.Variable(tf.random_normal([2048, 1024])),
'fc2': tf.Variable(tf.random_normal([1024, 1024])),
'out': tf.Variable(tf.random_normal([1024, 30])),
}

biases = {
'bc1_1': tf.Variable(tf.random_normal([8])),
'bc1_2': tf.Variable(tf.random_normal([8])),
'bc1_3': tf.Variable(tf.random_normal([8])),

'fc1': tf.Variable(tf.random_normal([1024])),
'fc2': tf.Variable(tf.random_normal([1024])),
'out': tf.Variable(tf.random_normal([30]))
}   


# the parameter restore
batch_size = 128
learning_rate = 0.01
weightreg_factor = 0.001  
 # regularization on the weights

num_epochs = 100
momentum = 0.9

use_early_stopping = True
#lr_of_ep = lambda ep: self.learning_rate/(1+0.2*ep) 
logs_path = './eval/tensorflow_logs/example/'
if not os.path.exists('./eval/'+eval_prefix+'/'):
    os.makedirs(logs_path)


out = create_model(X0, weights, biases)
cost = tf.sqrt(tf.reduce_sum(tf.square(out - Y), 1))  
# define l2 loss   
regularizer = 0
for w in weights:
    regularizer += tf.nn.l2_loss(weights[w])


cost = tf.reduce_mean(cost + weightreg_factor * regularizer)
#optimizer = tf.train.MomentumOptimizer(learning_rate=learning_rate, momentum = momentum)
optimizer =  tf.train.RMSPropOptimizer(learning_rate=learning_rate,decay=0.9,epsilon=1.0/100.)

train_op = optimizer.minimize(cost)

init = tf.global_variables_initializer()

# Create a summary to monitor cost tensor
tf.summary.scalar("loss", cost)
# Merge all summaries into a single op
merged_summary_op = tf.summary.merge_all()


with tf.Session() as sess:
    #Run the initializer
    sess.run(init)
    data_size = (train_data).shape[0]
    num_batches_per_epoch = int( data_size / batch_size) + 1

    # op to write logs to Tensorboard
    summary_writer = tf.summary.FileWriter(logs_path, graph=tf.get_default_graph())

    for epoch in range(num_epochs):
        for batch_num in range(num_batches_per_epoch):
            start_index = batch_num * batch_size
            end_index = min((batch_num + 1)* batch_size, data_size)
            data0 = train_data[start_index: end_index]
            label = train_gt3D_embed[start_index: end_index]
            learning_rate=learning_rate/(1+epoch*0.1)
            _, loss, summary, output = sess.run([train_op, cost, merged_summary_op, out], feed_dict={X0: data0, Y: label})

            summary_writer.add_summary(summary, epoch * num_batches_per_epoch + batch_num)
            print("batch_num " + str(batch_num) + ", Minibatch Loss= " + \
              "{:.4f}".format(loss) )
            
            print("the first output is {}".format(output[0,:]) + '/n' + "the true label is {}".format(label[0,:]))

**Comment by sinewy333** [comment]
Except for that, no changes have been made to the other code.and the data that  feed to the network is train_data, train_gt3D = trainDataSet.imgStackDepthOnly('train')，I'm sorry to bother you, but I can't find anything wrong. I hope you can help me.，Tank you very much。

**Comment by sinewy333** [comment]
is  the data used for training is "depth_1" in NYU dataset. ? or "synthdepth"?when I used the "depth_1",I found that the code didn't detect the position of the hand very well.when I used the "synthdepth"，the code detected  the position of the hand well.Thank you very much.

**Comment by moberweger** [comment]
Yes, that is correct. The data used for training is the "depth". The "
synthdepth" is simply a rendering of a 3d hand model. Thus it does not work
well for real camera data.

**Comment by sinewy333** [comment]
The code of  hand detector is based on gtorig [13]?which is from the data of label.But there is no such data in practice.How do we cut out the pictures with only hands?Thank you very much.

**Comment by moberweger** [comment]
If I understand your question correctly, we use gtorig[13] for training the
localizer. During testing, we first use the center of mass for hand
detection and the trained localizer for refining this location.

**Comment by sinewy333** [comment]
is used the main_nyu_com_refine for for refining the location of the hand？then use the main_nyu_posereg_embedding for training the  location of the joints，finally use the ORRef  for refine the   location of the joints？

**Comment by moberweger** [comment]
Yes, main_nyu_com_refine is for refining the location, and
main_nyu_posereg_embedding for pose prediction. ORRef is not published in
this repository, but you can optionally add it yourself.

**Comment by sinewy333** [comment]
I understand. Thanks a lot！

**Comment by sinewy333** [comment]
The test results of the model I trained came out. The test results in test 2 were much worse than those in test 1.Is it because the two data people are different?Does the size of the palm of different people affect the accuracy of the test? the result of your report is frome test1 or test2? look forward to your favourable reply. Thank you!

**Comment by sinewy333** [comment]
distance(mm) | train | test1 | test2

10                   | 25.84% | 15.82% | 0.12%
20                  | 74.57% | 52.01% | 11.44%
30                  | 88.89% | 73.44% | 32.71%
40                  | 94.54% | 82.38% | 48.61%
50                  | 97.43% | 89.18% | 64.52%
60                  | 98.85% | 93.81% | 76.74%
70                  | 99.43% | 97.01% | 85.31%
80                  | 99.76% | 98.44% | 90.31%

This is the result of my test（the  fraction of frames Within  distance(max)）

**Comment by moberweger** [comment]
You are correct that the results for test2 are worse. This is due to the fact that test2 is a different user with different hand size than the training user. Therefore, it is encouraged to adjust the crop size of the hand cube accordingly. The evaluation in the report is from the joint set of test1+test2.
